src/wrappers/digift/DigiftWrapper.sol

```diff
@@ -0,0 +1,1118 @@
+
+        _spendAllowance(msg.sender, address(this), shares);
+        _transfer(msg.sender, address(this), shares);
+
+        emit RedeemRequest(controller, owner, REQUEST_ID, msg.sender, shares);
+        return REQUEST_ID;
+    }
+
+    /**
+     * @notice Settle redemption requests for multiple nodes
+     * @dev Verifies Digift redemption event and distributes assets/shares proportionally
+     * @param nodes Array of node addresses to settle
+     * @param verifyArgs Offchain arguments for event verification
+     */
+    function settleRedeem(address[] calldata nodes, DigiftEventVerifier.OffchainArgs calldata verifyArgs)
+        external
+        nonReentrant
+        onlyManager
+    {
+        // Verify the Digift redemption event and get shares/assets amounts
+        (uint256 shares, uint256 assets) = digiftEventVerifier.verifySettlementEvent(
+            verifyArgs,
+            DigiftEventVerifier.OnchainArgs(
+                DigiftEventVerifier.EventType.REDEEM, address(subRedManagement), stToken, asset
+            )
+        );
+
+        SettleRedeemVars memory vars;
+        vars.globalPendingRedeemRequest = _globalState.pendingRedeemRequest;
+        require(vars.globalPendingRedeemRequest > 0, NothingToSettle());
+
+        // Process each node's redemption request
+        for (uint256 i; i < nodes.length; i++) {
+            NodeState storage node = _nodeState[nodes[i]];
+
+            uint256 nodePendingRedeemRequest = node.pendingRedeemRequest;
+
+            // Calculate proportional assets and shares for this node
+            uint256 assetsToReturn = nodePendingRedeemRequest.mulDiv(assets, vars.globalPendingRedeemRequest);
+            uint256 sharesToReimburse = nodePendingRedeemRequest.mulDiv(shares, vars.globalPendingRedeemRequest);
+
+            // Track totals for validation
+            vars.totalPendingRedeemRequestCheck += nodePendingRedeemRequest;
+            vars.totalAssetsToReturn += assetsToReturn;
+            vars.totalSharesToReimburse += sharesToReimburse;
+
+            // Handle dust accumulation on the last node
+            if (i == nodes.length - 1) {
+                if (vars.totalAssetsToReturn < assets || vars.totalSharesToReimburse < shares) {
+                    assetsToReturn += assets - vars.totalAssetsToReturn;
+                    sharesToReimburse += shares - vars.totalSharesToReimburse;
+                }
+            }
+
+            // Update node state
+            node.claimableRedeemRequest = nodePendingRedeemRequest;
+            node.pendingRedeemRequest = 0;
+            node.maxWithdraw = assetsToReturn;
+            node.pendingRedeemReimbursement = sharesToReimburse;
+
+            emit RedeemSettled(nodes[i], sharesToReimburse, assetsToReturn);
+        }
+
+        // Validate that all nodes have been processed
+        require(vars.totalPendingRedeemRequestCheck == vars.globalPendingRedeemRequest, NotAllNodesSettled());
```

Comment:
It’s also not ideal. For instance, there might be a situation where two nodes have already made requests, then forwardRequestsToDigift is called, and afterward, a third node submits a request (all nodes having the same amount - 100 shares). During settlement, instead of processing the first two nodes, nodes 1 and 3 might be processed instead, meaning node 3 is fulfilled earlier. This doesn’t introduce any vulnerability, but it results in a slightly unfair settlement due to the stToken value appreciation. Since such cases are highly unlikely and the effect is negligible, this tradeoff is considered acceptable.